synapse/lib/ast.py

```diff
@@ -219,9 +183,144 @@ def cancel(self):
+
+        'degrees': 1,
+
+        'filters': [
+            '-(#foo or #bar)',
+            '-(foo:bar or baz:faz)',
+        ],
+
+        'pivots': [
+            '-> * | limit 100',
+            '<- * | limit 100',
+        ]
+
+        'forms': {
+
+            'inet:fqdn':{
+                'filters': [],
+                'pivots': [],
+            }
+
+            '*': {
+                'filters': [],
+                'pivots': [],
+            },
+        },
+    }
+
+    # nodes which were original seeds have path.meta('graph:seed')
+    # all nodes have path.meta('edges') which is a list of (iden, info) tuples.
+    '''
+
+    def __init__(self, rules):
+
+        self.omits = {}
+        self.rules = rules
+
+        self.rules.setdefault('forms', {})
+        self.rules.setdefault('pivots', ())
+        self.rules.setdefault('filters', ())
+        self.rules.setdefault('degrees', 1)
+
+    async def omit(self, node):
+
+        answ = self.omits.get(node.buid)
+        if answ is not None:
+            return answ
+
+        for filt in self.rules.get('filters'):
+            if await node.filter(filt, user=self.user):
+                self.omits[node.buid] = True
+                return True
+
+        rules = self.rules['forms'].get(node.form.name)
+        if rules is None:
+            rules = self.rules['forms'].get('*')
+
+        if rules is None:
+            self.omits[node.buid] = False
+            return False
+
+        for filt in rules.get('filters', ()):
+            if await node.filter(filt, user=self.user):
+                self.omits[node.buid] = True
+                return True
+
+        self.omits[node.buid] = False
+        return False
+
+    async def pivots(self, node):
+
+        for pivq in self.rules.get('pivots'):
+
+            async for pivo in node.storm(pivq, user=self.user):
+                yield pivo
+
+        rules = self.rules['forms'].get(node.form.name)
+        if rules is None:
+            rules = self.rules['forms'].get('*')
+
+        if rules is None:
+            return
+
+        for pivq in rules.get('pivots', ()):
+            async for pivo in node.storm(pivq, user=self.user):
+                yield pivo
+
+    async def run(self, runt, genr):
+
+        done = {}
+        degrees = self.rules.get('degrees')
+
+        self.user = runt.user
+
+        async for node, path in genr:
+
+            if await self.omit(node):
+                continue
+
+            path.meta('graph:seed', True)
+
+            todo = collections.deque([(node, path, 0)])
+
+            while todo:
+
+                tnode, tpath, tdist = todo.popleft()
+
+                # filter out nodes that we've already done at
+                # the given distance or less... (best possible)
+                donedist = done.get(tnode.buid)
+                if donedist is not None and donedist <= tdist:
+                    continue
+
+                done[tnode.buid] = tdist
+
+                edges = {}
```

Comment:
you don't like sets?

Reply:
ah.  yea, that one could have been a set.  i'll tweak. :)